### PR TITLE
Support for System and User Assigned Managed Identities in `azurerm_eventgrid_domain`, `azurerm_eventgrid_topic` and `azurerm_eventgrid_system_topic`

### PR DESCRIPTION
**Pull request by jrauschenbusch**
Reopened PR for https://github.com/hashicorp/terraform-provider-azurerm/pull/9906

Issue: https://github.com/hashicorp/terraform-provider-azurerm/issues/9525

@mbfrahry Please review

### CODE REVIEW

File: website/docs/r/eventgrid_topic.html.markdown [hunk 1]
```diff
@@ -56,6 +58,16 @@ The following arguments are supported:
 
 ---
 
+A `identity` block supports the following:
+
+* `type` - Specifies the identity type of Event Grid Domain. Possible values are `SystemAssigned` (where Azure will generate a Principal for you) or `UserAssigned` where you can specify the User Assigned Managed Identity IDs in the `identity_ids` field.
```
Comment (stephybun) [inline website/docs/r/eventgrid_topic.html.markdown:63]: This should be 
```suggestion
* `type` - Specifies the identity type of Event Grid Topic. Possible values are `SystemAssigned` (where Azure will generate a Principal for you) or `UserAssigned` where you can specify the User Assigned Managed Identity IDs in the `identity_ids` field.
```

File: website/docs/r/eventgrid_topic.html.markdown [hunk 2]
```diff
@@ -56,6 +58,16 @@ The following arguments are supported:
...
+
+~> **NOTE:** When `type` is set to `SystemAssigned`, The assigned `principal_id` and `tenant_id` can be retrieved after the Event Grid Domain has been created. More details are available below.
```
Comment (stephybun) [inline website/docs/r/eventgrid_topic.html.markdown:65]: Same here 
```suggestion
~> **NOTE:** When `type` is set to `SystemAssigned`, The assigned `principal_id` and `tenant_id` can be retrieved after the Event Grid Topic has been created. More details are available below.
```

File: website/docs/r/eventgrid_topic.html.markdown [hunk 3]
```diff
@@ -100,9 +112,23 @@ The following attributes are exported:
 
 * `secondary_access_key` - The Secondary Shared Access Key associated with the EventGrid Topic.
 
-## Timeouts
+* `identity` - An `identity` block as defined below, which contains the Managed Service Identity information for this Event Grid Domain.
```
Comment (stephybun) [inline website/docs/r/eventgrid_topic.html.markdown:115]: Same here 
```suggestion
* `identity` - An `identity` block as defined below, which contains the Managed Service Identity information for this Event Grid Topic.
```

File: website/docs/r/eventgrid_topic.html.markdown [hunk 4]
```diff
@@ -100,9 +112,23 @@ The following attributes are exported:
...
 
+---
 
+A `identity` block exports the following:
```
Comment (stephybun) [inline website/docs/r/eventgrid_topic.html.markdown:119]: Going by the doc comments from #9906 this should be 
```suggestion
A `identity` block supports the following:
```

File: website/docs/r/eventgrid_topic.html.markdown [hunk 5]
```diff
@@ -100,9 +112,23 @@ The following attributes are exported:
...
+
+* `type` - Specifies the type of Managed Service Identity that is configured on this Event Grid Domain.
```
Comment (stephybun) [inline website/docs/r/eventgrid_topic.html.markdown:121]: Same here
```suggestion
* `type` - Specifies the type of Managed Service Identity that is configured on this Event Grid Topic.
```

File: website/docs/r/eventgrid_topic.html.markdown [hunk 6]
```diff
@@ -100,9 +112,23 @@ The following attributes are exported:
...
+
+* `principal_id` - Specifies the Principal ID of the System Assigned Managed Service Identity that is configured on this Event Grid Domain.
```
Comment (stephybun) [inline website/docs/r/eventgrid_topic.html.markdown:123]:
```suggestion
* `principal_id` - Specifies the Principal ID of the System Assigned Managed Service Identity that is configured on this Event Grid Topic.
```

File: website/docs/r/eventgrid_topic.html.markdown [hunk 7]
```diff
@@ -100,9 +112,23 @@ The following attributes are exported:
...
+
+* `tenant_id` - Specifies the Tenant ID of the System Assigned Managed Service Identity that is configured on this Event Grid Domain.
```
Comment (stephybun) [inline website/docs/r/eventgrid_topic.html.markdown:125]:
```suggestion
* `tenant_id` - Specifies the Tenant ID of the System Assigned Managed Service Identity that is configured on this Event Grid Topic.
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks for reopening this PR and for fixing most of the comments left from #9906 @jrauschenbusch. The tests are passing, once the comments left on on the docs are fixed then this LGTM 👍🏼.

**Comment by jrauschenbusch** [comment]
@stephybun Done

**Review by katbyte — APPROVED** [review]
Thanks both! LGTM 🚀

**Comment by katbyte** [comment]
@jrauschenbusch - this is ready to go once the merge conflicts are resolved!

**Comment by jrauschenbusch** [comment]
@katbyte Could you please fix it? I'm currently in vacation with mobile phone access only.

**Comment by javier-moreno-h** [comment]
Thanks a lot for reopening this PR for fixing #9925. What about **azurerm_eventgrid_system_topic** ? Will it also support identities ? King Regards.

**Review by stephybun — DISMISSED** [review]

**Comment by jrauschenbusch** [comment]
@javier-moreno-h Until now this PR just comprised the identity support for `azurerm_eventgrid_domain` and `azurerm_eventgrid_topic`.  I've now added also support for `azurerm_eventgrid_system_topic`. I'll do a few tests and give an update here if the state is mergable. Basically it was the really same code changes as needed for the topic resource. Hence a further PR review should be just a comparison of those two resources.

**Comment by stephybun** [comment]
@jrauschenbusch please re-request review when ready.

**Comment by javier-moreno-h** [comment]
> @javier-moreno-h Until now this PR just comprised the identity support for `azurerm_eventgrid_domain` and `azurerm_eventgrid_topic`. I've now added also support for `azurerm_eventgrid_system_topic`. I'll do a few tests and give an update here if the state is mergable. Basically it was the really same code changes as needed for the topic resource. Hence a further PR review should be just a comparison of those two resources.

@jrauschenbusch thanks a lot. I really appreciate it. I will be watching this thread. Regards.

**Comment by jrauschenbusch** [comment]
@stephybun PR  is now ready to be re-reviewed. I had to adjust the expand/flatten function names due to collisions with another PR which was merged before. Basically there is some misalignment of the identity implementation in the Azure SDK. Therefore we need multiple functions for this job within the event grid module. This was already addressed by @tombuildsstuff within the azure-rest-api-specs repository (see https://github.com/Azure/azure-rest-api-specs/issues/13340). Also i had to adjust a few parts of the system topic tests, due to source resource restrictions (allowed lengths of storage accounts).

Below you'll find the output of my acceptance test execution. I ran the whole event grid acceptance test collection.

<details>
<summary>Console Output</summary>

```
make acctests SERVICE='eventgrid' TESTARGS='-run=TestAccEventGrid' TESTTIMEOUT='60m' 
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/eventgrid -run=TestAccEventGrid -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccEventGridDomainDataSource_basic
=== PAUSE TestAccEventGridDomainDataSource_basic
=== RUN   TestAccEventGridDomain_basic
=== PAUSE TestAccEventGridDomain_basic
=== RUN   TestAccEventGridDomain_requiresImport
=== PAUSE TestAccEventGridDomain_requiresImport
=== RUN   TestAccEventGridDomain_mapping
=== PAUSE TestAccEventGridDomain_mapping
=== RUN   TestAccEventGridDomain_basicWithTags
=== PAUSE TestAccEventGridDomain_basicWithTags
=== RUN   TestAccEventGridDomain_inboundIPRules
=== PAUSE TestAccEventGridDomain_inboundIPRules
=== RUN   TestAccEventGridDomain_basicWithSystemManagedIdentity
=== PAUSE TestAccEventGridDomain_basicWithSystemManagedIdentity
=== RUN   TestAccEventGridDomain_basicWithUserAssignedManagedIdentity
=== PAUSE TestAccEventGridDomain_basicWithUserAssignedManagedIdentity
=== RUN   TestAccEventGridDomainTopicDataSource_basic
=== PAUSE TestAccEventGridDomainTopicDataSource_basic
=== RUN   TestAccEventGridDomainTopic_basic
=== PAUSE TestAccEventGridDomainTopic_basic
=== RUN   TestAccEventGridDomainTopic_requiresImport
=== PAUSE TestAccEventGridDomainTopic_requiresImport
=== RUN   TestAccEventGridEventSubscription_basic
=== PAUSE TestAccEventGridEventSubscription_basic
=== RUN   TestAccEventGridEventSubscription_requiresImport
=== PAUSE TestAccEventGridEventSubscription_requiresImport
=== RUN   TestAccEventGridEventSubscription_eventHubID
=== PAUSE TestAccEventGridEventSubscription_eventHubID
=== RUN   TestAccEventGridEventSubscription_serviceBusQueueID
=== PAUSE TestAccEventGridEventSubscription_serviceBusQueueID
=== RUN   TestAccEventGridEventSubscription_serviceBusTopicID
=== PAUSE TestAccEventGridEventSubscription_serviceBusTopicID
=== RUN   TestAccEventGridEventSubscription_update
=== PAUSE TestAccEventGridEventSubscription_update
=== RUN   TestAccEventGridEventSubscription_filter
=== PAUSE TestAccEventGridEventSubscription_filter
=== RUN   TestAccEventGridEventSubscription_advancedFilter
=== PAUSE TestAccEventGridEventSubscription_advancedFilter
=== RUN   TestAccEventGridEventSubscription_advancedFilterMaxItems
=== PAUSE TestAccEventGridEventSubscription_advancedFilterMaxItems
=== RUN   TestAccEventGridEventSubscription_identity
=== PAUSE TestAccEventGridEventSubscription_identity
=== RUN   TestAccEventGridSystemTopicEventSubscription_basic
=== PAUSE TestAccEventGridSystemTopicEventSubscription_basic
=== RUN   TestAccEventGridSystemTopicEventSubscription_requiresImport
=== PAUSE TestAccEventGridSystemTopicEventSubscription_requiresImport
=== RUN   TestAccEventGridSystemTopicEventSubscription_eventHubID
=== PAUSE TestAccEventGridSystemTopicEventSubscription_eventHubID
=== RUN   TestAccEventGridSystemTopicEventSubscription_serviceBusQueueID
=== PAUSE TestAccEventGridSystemTopicEventSubscription_serviceBusQueueID
=== RUN   TestAccEventGridSystemTopicEventSubscription_serviceBusTopicID
=== PAUSE TestAccEventGridSystemTopicEventSubscription_serviceBusTopicID
=== RUN   TestAccEventGridSystemTopicEventSubscription_update
=== PAUSE TestAccEventGridSystemTopicEventSubscription_update
=== RUN   TestAccEventGridSystemTopicEventSubscription_filter
=== PAUSE TestAccEventGridSystemTopicEventSubscription_filter
=== RUN   TestAccEventGridSystemTopicEventSubscription_advancedFilter
=== PAUSE TestAccEventGridSystemTopicEventSubscription_advancedFilter
=== RUN   TestAccEventGridSystemTopicEventSubscription_advancedFilterMaxItems
=== PAUSE TestAccEventGridSystemTopicEventSubscription_advancedFilterMaxItems
=== RUN   TestAccEventGridSystemTopicEventSubscription_identity
=== PAUSE TestAccEventGridSystemTopicEventSubscription_identity
=== RUN   TestAccEventGridSystemTopic_basic
=== PAUSE TestAccEventGridSystemTopic_basic
=== RUN   TestAccEventGridSystemTopic_policyStates
=== PAUSE TestAccEventGridSystemTopic_policyStates
=== RUN   TestAccEventGridSystemTopic_requiresImport
=== PAUSE TestAccEventGridSystemTopic_requiresImport
=== RUN   TestAccEventGridSystemTopic_complete
=== PAUSE TestAccEventGridSystemTopic_complete
=== RUN   TestAccEventGridSystemTopic_basicWithSystemManagedIdentity
=== PAUSE TestAccEventGridSystemTopic_basicWithSystemManagedIdentity
=== RUN   TestAccEventGridSystemTopic_basicWithUserAssignedManagedIdentity
=== PAUSE TestAccEventGridSystemTopic_basicWithUserAssignedManagedIdentity
=== RUN   TestAccEventGridTopicDataSource_basic
=== PAUSE TestAccEventGridTopicDataSource_basic
=== RUN   TestAccEventGridTopic_basic
=== PAUSE TestAccEventGridTopic_basic
=== RUN   TestAccEventGridTopic_requiresImport
=== PAUSE TestAccEventGridTopic_requiresImport
=== RUN   TestAccEventGridTopic_mapping
=== PAUSE TestAccEventGridTopic_mapping
=== RUN   TestAccEventGridTopic_basicWithTags
=== PAUSE TestAccEventGridTopic_basicWithTags
=== RUN   TestAccEventGridTopic_inboundIPRules
=== PAUSE TestAccEventGridTopic_inboundIPRules
=== RUN   TestAccEventGridTopic_basicWithSystemManagedIdentity
=== PAUSE TestAccEventGridTopic_basicWithSystemManagedIdentity
=== RUN   TestAccEventGridTopic_basicWithUserAssignedManagedIdentity
=== PAUSE TestAccEventGridTopic_basicWithUserAssignedManagedIdentity
=== CONT  TestAccEventGridDomainDataSource_basic
=== CONT  TestAccEventGridSystemTopicEventSubscription_eventHubID
=== CONT  TestAccEventGridEventSubscription_requiresImport
=== CONT  TestAccEventGridSystemTopicEventSubscription_requiresImport
=== CONT  TestAccEventGridSystemTopicEventSubscription_basic
=== CONT  TestAccEventGridEventSubscription_identity
=== CONT  TestAccEventGridEventSubscription_advancedFilterMaxItems
=== CONT  TestAccEventGridEventSubscription_advancedFilter
=== CONT  TestAccEventGridEventSubscription_filter
=== CONT  TestAccEventGridEventSubscription_update
=== CONT  TestAccEventGridEventSubscription_serviceBusTopicID
=== CONT  TestAccEventGridEventSubscription_serviceBusQueueID
--- PASS: TestAccEventGridEventSubscription_advancedFilterMaxItems (131.63s)
=== CONT  TestAccEventGridSystemTopic_basicWithSystemManagedIdentity
--- PASS: TestAccEventGridEventSubscription_advancedFilter (131.92s)
=== CONT  TestAccEventGridEventSubscription_eventHubID
--- PASS: TestAccEventGridEventSubscription_filter (138.49s)
=== CONT  TestAccEventGridDomain_basicWithSystemManagedIdentity
--- PASS: TestAccEventGridDomainDataSource_basic (149.60s)
=== CONT  TestAccEventGridTopic_basicWithUserAssignedManagedIdentity
--- PASS: TestAccEventGridSystemTopicEventSubscription_basic (152.47s)
=== CONT  TestAccEventGridEventSubscription_basic
--- PASS: TestAccEventGridSystemTopicEventSubscription_requiresImport (166.92s)
=== CONT  TestAccEventGridTopic_basicWithSystemManagedIdentity
--- PASS: TestAccEventGridEventSubscription_requiresImport (189.39s)
=== CONT  TestAccEventGridDomainTopic_requiresImport
--- PASS: TestAccEventGridEventSubscription_update (240.98s)
=== CONT  TestAccEventGridTopic_inboundIPRules
--- PASS: TestAccEventGridTopic_basicWithUserAssignedManagedIdentity (100.95s)
=== CONT  TestAccEventGridDomainTopic_basic
--- PASS: TestAccEventGridSystemTopic_basicWithSystemManagedIdentity (124.37s)
=== CONT  TestAccEventGridDomainTopicDataSource_basic
--- PASS: TestAccEventGridTopic_basicWithSystemManagedIdentity (97.10s)
=== CONT  TestAccEventGridDomain_basicWithUserAssignedManagedIdentity
--- PASS: TestAccEventGridSystemTopicEventSubscription_eventHubID (284.47s)
=== CONT  TestAccEventGridTopic_basicWithTags
=== CONT  TestAccEventGridTopic_basic
--- PASS: TestAccEventGridEventSubscription_identity (284.60s)
--- PASS: TestAccEventGridEventSubscription_basic (134.24s)
=== CONT  TestAccEventGridTopicDataSource_basic
--- PASS: TestAccEventGridEventSubscription_serviceBusTopicID (292.68s)
=== CONT  TestAccEventGridTopic_mapping
--- PASS: TestAccEventGridDomain_basicWithSystemManagedIdentity (160.88s)
=== CONT  TestAccEventGridSystemTopic_basicWithUserAssignedManagedIdentity
--- PASS: TestAccEventGridEventSubscription_serviceBusQueueID (311.46s)
=== CONT  TestAccEventGridTopic_requiresImport
--- PASS: TestAccEventGridDomain_basicWithUserAssignedManagedIdentity (102.08s)
=== CONT  TestAccEventGridDomain_mapping
--- PASS: TestAccEventGridDomainTopic_requiresImport (185.88s)
=== CONT  TestAccEventGridDomain_inboundIPRules
--- PASS: TestAccEventGridTopic_basicWithTags (98.64s)
=== CONT  TestAccEventGridDomain_basicWithTags
--- PASS: TestAccEventGridTopic_mapping (98.45s)
=== CONT  TestAccEventGridSystemTopicEventSubscription_update
--- PASS: TestAccEventGridTopic_inboundIPRules (158.30s)
=== CONT  TestAccEventGridSystemTopic_policyStates
--- PASS: TestAccEventGridEventSubscription_eventHubID (281.64s)
=== CONT  TestAccEventGridSystemTopicEventSubscription_advancedFilter
--- PASS: TestAccEventGridSystemTopic_basicWithUserAssignedManagedIdentity (126.34s)
=== CONT  TestAccEventGridSystemTopic_complete
--- PASS: TestAccEventGridDomainTopic_basic (180.45s)
=== CONT  TestAccEventGridSystemTopicEventSubscription_filter
--- PASS: TestAccEventGridDomainTopicDataSource_basic (175.96s)
=== CONT  TestAccEventGridSystemTopic_requiresImport
--- PASS: TestAccEventGridTopicDataSource_basic (165.40s)
=== CONT  TestAccEventGridSystemTopicEventSubscription_serviceBusTopicID
--- PASS: TestAccEventGridTopic_basic (183.44s)
=== CONT  TestAccEventGridSystemTopicEventSubscription_serviceBusQueueID
--- PASS: TestAccEventGridDomain_inboundIPRules (98.33s)
=== CONT  TestAccEventGridDomain_requiresImport
--- PASS: TestAccEventGridDomain_basicWithTags (100.91s)
=== CONT  TestAccEventGridSystemTopic_basic
--- PASS: TestAccEventGridTopic_requiresImport (197.18s)
=== CONT  TestAccEventGridSystemTopicEventSubscription_identity
--- PASS: TestAccEventGridDomain_mapping (159.34s)
=== CONT  TestAccEventGridDomain_basic
--- PASS: TestAccEventGridSystemTopic_complete (115.32s)
=== CONT  TestAccEventGridSystemTopicEventSubscription_advancedFilterMaxItems
--- PASS: TestAccEventGridSystemTopic_policyStates (154.36s)
--- PASS: TestAccEventGridSystemTopic_requiresImport (127.01s)
--- PASS: TestAccEventGridSystemTopicEventSubscription_filter (146.12s)
--- PASS: TestAccEventGridSystemTopicEventSubscription_update (193.78s)
--- PASS: TestAccEventGridSystemTopic_basic (112.54s)
--- PASS: TestAccEventGridDomain_requiresImport (171.41s)
--- PASS: TestAccEventGridSystemTopicEventSubscription_advancedFilter (234.56s)
--- PASS: TestAccEventGridSystemTopicEventSubscription_advancedFilterMaxItems (141.83s)
--- PASS: TestAccEventGridDomain_basic (159.84s)
--- PASS: TestAccEventGridSystemTopicEventSubscription_serviceBusQueueID (248.56s)
--- PASS: TestAccEventGridSystemTopicEventSubscription_serviceBusTopicID (282.77s)
--- PASS: TestAccEventGridSystemTopicEventSubscription_identity (339.07s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/eventgrid     849.227s
```
</details>

**Review by stephybun — APPROVED** [review]
Thanks @jrauschenbusch LGTM 🥳

**Pull request merged**